adapters/repos/db/shard_combine_multi_target.go

```diff
@@ -0,0 +1,278 @@
+					return nil, nil, fmt.Errorf("number of weights in join does not match number of results")
+				}
+				if !ok {
+					tmp = objAndDistance{obj: results[i][j], distance: 0}
+				}
+
+				weight := multiTargetCombination.Weights[targetVectors[i]]
+				tmp.distance += weight * dists[i][j]
+			}
+			combinedResults[uid] = tmp
+
+		}
+		collectMissingIds(localIDs, missingIDs, targetVectors, searchVectors, i)
+	}
+	if multiTargetCombination.Type != dto.Minimum {
+		if err := getScoresOfMissingResults(ctx, shard, logger, missingIDs, &ResultContainerStandard{combinedResults}, multiTargetCombination.Weights, allIDs); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	queue := priorityqueue.NewMin[float32](limit)
+	uuidCounter := make([]strfmt.UUID, len(combinedResults))
+	count := uint64(0)
+	for id, res := range combinedResults {
+		if targetDist > 0 && res.distance > targetDist { // targetDist == 0 means no target distance is set
+			continue
+		}
+		uuidCounter[count] = id
+		queue.Insert(count, res.distance)
+		count++
+
+	}
+	returnResults := make([]*storobj.Object, 0, queue.Len())
+	returnDists := make([]float32, 0, queue.Len())
+	lim := queue.Len()
+	for i := 0; i < lim; i++ {
+		item := queue.Pop()
+		returnResults = append(returnResults, combinedResults[uuidCounter[item.ID]].obj)
+		returnDists = append(returnDists, combinedResults[uuidCounter[item.ID]].distance)
+	}
+
+	if limit > len(returnResults) {
+		limit = len(returnResults)
+	}
+
+	return returnResults[:limit], returnDists[:limit], nil
+}
+
+func collectMissingIds(localIDs map[strfmt.UUID]*storobj.Object, missingIDs map[strfmt.UUID]targetVectorData, targetVectors []string, searchVectors [][]float32, i int) {
+	for id := range localIDs {
+		val, ok := missingIDs[id]
+		if !ok {
+			val = targetVectorData{target: []string{targetVectors[i]}, searchVector: [][]float32{searchVectors[i]}}
+		} else {
+			val.target = append(val.target, targetVectors[i])
+			val.searchVector = append(val.searchVector, searchVectors[i])
+		}
+		missingIDs[id] = val
+	}
+}
+
+func getScoresOfMissingResults(ctx context.Context, shard DistanceForVector, logger logrus.FieldLogger, missingIDs map[strfmt.UUID]targetVectorData, combinedResults ResultContainer, weights map[string]float32, allIDs map[strfmt.UUID]*storobj.Object) error {
+	if len(missingIDs) == 0 {
+		return nil
+	}
+
+	eg, ctx := enterrors.NewErrorGroupWithContextWrapper(logger, ctx)
+	eg.SetLimit(_NUMCPU * 2)
+	mutex := sync.Mutex{}
+	for id, targets := range missingIDs {
+		id := id
+		targets := targets
+		f := func() error {
+			distances, err := shard.VectorDistanceForQuery(ctx, allIDs[id].ID(), targets.searchVector, targets.target)
+			mutex.Lock()
+			if err != nil {
+				// when we cannot look up missing distances for an object, it will be removed from the result list
+				combinedResults.RemoveIdFromResult(id)
+			} else {
+				combinedResults.AddScores(id, targets.target, distances, weights)
+			}
+			mutex.Unlock()
```

Comment:
maybe safe would be to put the `mutex.Unlock()` to line 263 with `defer mutex.Unlock()`? we are sure that nothing can go wrong during `combinedResults.RemoveIdFromResult` or `combinedResults.AddScores`?

Reply:
they are very simple, but using defer doesn't hurt